[sw/crypto] Implement SCA guidelines from OT docs for p256_base.s and…

pqcfox opened the pull request:
This PR implements the official SCA guidelines for OTBN for `p256_base.s` and `p256_sign.s`, as listed in the [Secure Coding for Cryptography section](https://opentitan.org/book/doc/contributing/style_guides/otbn_style_guide.html#secure-coding-for-cryptography) of the OTBN Assembly Style Guide.

Performance overhead is +0.48% latency for ECDSA signing, and +<0.01% latency for ECDSA verification.

(This PR concludes my SCA pass for P256, since the remaining files don't act on secrets which need SCA protection.)

pqcfox commented:
Apologies for delay, just saw the CI failures. Fixing ASAP tomorrow morning PT.

jadephilipoom commented on sw/otbn/crypto/p256_base.s:
```diff
@@ -151,59 +151,95 @@ p256_reduce:
+     additions, and shift the result another bit to the right. */
 
-  /* w25 = q1[256] = x[511:256] >> 255 */
+  /* w25 = q1[256] = a[511:256] >> 255 */
```
In this case I was using `x` to match the HAC reference -- I'm fine with changing it to `a`, seems equally valid, but then I'd want to change line 121 to match :slightly_smiling_face:

pqcfox replied:
Ooh, good catch, thank you! I reverted everything to `a` and changed the variable name in the header comment from `x` -> `a` as well.

jadephilipoom commented on sw/otbn/crypto/p256_base.s:
```diff
@@ -151,59 +151,95 @@ p256_reduce:
+
+    N.B. The bn.mulqacc.so.z instruction below is to clear the value of ACC as
+    well as the flags set by the prior mulqacc.so instructions, as p256_reduce
+    may be called on secret shares of the same value back-to-back. For instance,
+    p256_sign calls mod_mul_32x128 to compute d0 * alpha and  d1 * alpha
+    back-to-back, where alpha is a random multiplicative value for masking and
+    d0, d1 are additive shares of the ECDSA private key d.
```
nit: I'd maybe leave out specific reference to `p256_sign` and `mod_mul`? It's the kind of comment that can easily become out of sync with the code it references, e.g. if we changed the name of alpha or something. It would still be fine to say that we might call it in this pattern to do multiplicative masking in general.

pqcfox replied:
Removed the specific name references and recommented--thank you!

jadephilipoom commented on sw/otbn/crypto/p256_base.s:
```suggestion
    bn.wsrr   w7, URND
```
SInce this is for masking, we can use the internal PRNG instead of waiting for guaranteed key-quality randomness. Same for the other share.

pqcfox replied:
Added!

jadephilipoom commented on sw/otbn/crypto/p256_base.s:
Thinking about this some more, maybe it would work to just shift out the low bits instead and do the select based on `L` instead of `M̀`? Same number of instructions, but then every bit except the target bit would be random instead of 8 bits being untouched.
```
bn.wsrr   w26, URND
bn.rshi    w7, w26, w1 >> 255
```

pqcfox replied:
Added this--this was perfect, thank you so much!

jadephilipoom commented on sw/otbn/crypto/p256_base.s:
```diff
@@ -1834,9 +1912,17 @@ p256_key_from_seed:
        [w11,w10] <= x1 mod (n << 64) = d1 */
   bn.sel    w10, w10, w24, FG0.C
   bn.sel    w11, w11, w25, FG0.C
+  bn.sub    w23, w23, w23
```
nit: add a "dummy instruction" comment here maybe?

pqcfox replied:
Added!

pqcfox commented:
Thanks for excellent review! Addressed the above comments :)

Wasn't able to reproduce the test failures from previous CI run locally, so going to let it run here again in case the CI failures were a fluke.

pqcfox commented:
Update; I don't know quite what happened, but I think the wrong commit pushed--resolving that now.

pqcfox commented:
Modified the above to match above suggestions; did require adding one instruction to shift the other share's MSb into the LSb of a word, but the only performance change is making the PR +0.56% latency overhead relative to `master` (instead of the previous +0.48%).

andrea-caforio reviewed (commented):
For a hardened double-and-add function I would prefer to individually compute `k0 * P` and `k1 * P` to simplify any additional hardening measures on top of it as it renders the implementation less opaque. It further makes it more amendable to FI hardening.

This ad-hoc approach of somehow re-masking `k0` before extracting the MSB of `k0 ^ k1` is a delicate choice; in general any kind of recombination of related shares in should be avoided in my opinion. It could be an option after having experimentally verified its soundness but with the lack of any leakage report I am hesitant with such optimizations.

As the code is already slow, can you explain why this optimized double-and-add strategy where only a single iteration is being computed is necessary? 

Also, applying this kind of band-aid hardening approach to P-384 where intermediate variables are constantly fetched and retrieved from DMEM introduces even more complexity that is rather unwarranted and makes the two curve implementations diverge even more from each other.

jadephilipoom commented:
> For a hardened double-and-add function I would prefer to individually compute `k0 * P` and `k1 * P` to simplify any additional hardening measures on top of it as it renders the implementation less opaque. It further makes it more amendable to FI hardening.

The objective is to eventually compute `(k0 + k1) * P`. But elliptic curve points are not integers and the `*` there is not integer multiplication. There is no way to multiply two points together, you can _only_ add. So once you have `k0 * P` and `k1 * P`, it's not straightforward to see how you would get to `(k0 + k1) * P`. Obviously it would be better to compute on both shares separately if possible, but with additive masking that isn't straightforward. This is why we use the extra redundant bits in the shares, so the attacker needs to learn ~64 bits before they get information about the actual key.

I have some ideas for more extensive changes we could do to make this code more SCA-safe, but that is well outside the scope of this PR which is about applying the coding guidelines, not redoing our entire P256 masking scheme.

jadephilipoom commented:
Okay, unless I've missed something there are no objections to merging this PR? I'll go ahead and hit the button soon unless someone tells me to hold off.
